### DISCUSSION TITLE
mTRL Comparison skrf vs NIST/TUG mTRL

### DISCUSSION THREAD

**Discussion by Zwelckovich** [scikit-rf/scikit-rf#805] [Q&A] [answered]
Hello,
using the skrf mTRL and the mTRL from Ziad Hatab (https://github.com/ZiadHatab/multiline-trl-calibration) i have different results. The skrf mTRL are not in the same range as the TUG mTRL. 
![Figure_1](https://user-images.githubusercontent.com/54358761/203278552-7a958275-0370-4859-b551-7bfe3c3e5155.png)

I attached the source files and the touchstone files in the zip file.
[skrf_issue.zip](https://github.com/scikit-rf/scikit-rf/files/10065975/skrf_issue.zip)

**Comment by Vinc0110** [comment]
@ZiadHatab

**Comment by ZiadHatab** [comment, marked as answer]
@Zwelckovich 

Hi, the mTRL in skrf is working fine. The problem is that you used a non-zero thru length (37.25mm). In my code, after calibration, the reference is set in the middle of the thru. With skrf, the reference plane is automatically shifted backward half the length of the thru you specify.

To get the same result as skrf, you can use the below function in my implementation before applying the calibration to the dut. I already have a section in my repo discussing this. Also regarding skrf, they already mention this in their doc https://scikit-rf.readthedocs.io/en/latest/api/calibration/generated/skrf.calibration.calibration.NISTMultilineTRL.html

```python
cal.shift_plane(-line_lengths[0]/2)
```

![image](https://user-images.githubusercontent.com/46247580/203404019-cf41093d-0691-4ef8-a789-5fdfafed1a14.png)

**Comment by Zwelckovich** [comment]
@ZiadHatab 
Thanks, the non-zero thru length was the problem. 
Another question: Is there a quick way to convert your 8(6) error terms to a 12 terms error? I analyzed your code but iam not sure if i can use the skrf convert_8term_2_12term way:
`Edf = coefs_8term['forward directivity']
    Esf = coefs_8term['forward source match']
    Erf = coefs_8term['forward reflection tracking']

    Edr = coefs_8term['reverse directivity']
    Esr = coefs_8term['reverse source match']
    Err = coefs_8term['reverse reflection tracking']

    gamma_f = coefs_8term['forward switch term']
    gamma_r = coefs_8term['reverse switch term']
    k = coefs_8term['k']
    k_first = coefs_8term.get('k first', k)
    k_second = coefs_8term.get('k second', k)`

I can get ERF,EDF,etc from your `cal.coefs`.  But iam not sure about the the `gamma_f ` and `gamma_r`. There is a `calc.switch` that could match. And the `k` because `calc.K` is a complex number in your case.

Thanks for your help.

**Reply by ZiadHatab** [reply]
Hi!

You are right, there is no conversion function to 12-term in my code, but you can compute them from the information you have. The error coefficients used in skrf (up to my knowledge) are based on the PowerPoint presentation "Network Analyzer Error Models and Calibration Methods" by Doug Rytting (Agilent). You can just google it and you will find a copy to download. In the last pages of that PowerPoint presentation, you find the equations relating 8-term and 10(12)-term model.  One important thing to note is that the k i'm using is completely different and has nothing to do with the k in skrf notation (check Doug Rytting PowerPoint slides).  If you want to know more about the k I use, you can check my PowerPoint slides: https://pure.tugraz.at/ws/portalfiles/portal/46207898/ziad_ARFTG_98_presentation.pdf 

Alternatively, I suggest that you take the equations directly from the below paper in Section VI ("Relationship between twelve-term and error-box parameters"):
R. B. Marks, "Formulations of the Basic Vector Network Analyzer Error Model including Switch-Terms," 50th ARFTG Conference Digest, 1997, pp. 115-126, doi: 10.1109/ARFTG.1997.327265. https://ieeexplore.ieee.org/document/4119948 

Maybe in the future I will include the conversion equations in my code. For now, I leave it to you to program the equations (you already have everything).

**Reply by Zwelckovich** [reply]
@ZiadHatab  
Thanks, i added the conversion to your mtrl.py. If you have some free time you can review the code and you are free to merge it to your repo. 
Greetings
Sven
[mTRL.zip](https://github.com/scikit-rf/scikit-rf/files/10112398/mTRL.zip)

**Reply by ZiadHatab** [reply]
Thanks. I updated my code: https://github.com/ZiadHatab/multiline-trl-calibration

**Comment by jhillairet** [comment]
OK, Thank you @ZiadHatab. 

I'll close the issue and will transform it as a discussion. Please feel free to continue the discussion as needed.
